src/osd/PrimaryLogPG.cc

```diff
@@ -8326,6 +8369,36 @@ int PrimaryLogPG::prepare_transaction(OpContext *ctx)
     return -EINVAL;
   }
 
+  if (ctx->obc->obs.oi.has_manifest() && ctx->obc->obs.oi.soid.snap == CEPH_NOSNAP) {
+    if ((ctx->obs->exists && !ctx->obs->oi.is_whiteout()) &&
+	ctx->snapc.snaps.size() &&
+	ctx->snapc.snaps[0] > ctx->new_snapset.seq) {
+      // increase refcount if manifest object is snapshotted and need to make a clone 
+      if (ctx->op_finishers[0] == nullptr) {
+	//fin->refs = ctx->obs->oi.manifest.chunk_map.size();
+	std::shared_ptr<int> refcount = std::make_shared<int>();
+	*refcount = ctx->obs->oi.manifest.chunk_map.size();
+
+	for (auto p : ctx->obs->oi.manifest.chunk_map) {
+	  if (p.second.has_reference()) {
+	    object_locator_t target_oloc(p.second.oid);
+	    hobject_t source_coid = ctx->obc->obs.oi.soid;
+	    source_coid.snap = ctx->snapc.seq;
+	    RefCountCallback *fin = new RefCountCallback(
+	      this, ctx, NULL, get_last_peering_reset());
+	    fin->refs = refcount;
+	    refcount_manifest(ctx->obc, source_coid, target_oloc, p.second.oid, SnapContext(),
+			      true, fin);
+	  }
+	}
+	ctx->op_finishers[0].reset(new SetManifestFinisher());
+	return -EINPROGRESS;
+      } else {
+	ctx->op_finishers.erase(0);
+      }
+    }
+  }
+
```

Comment:
This part of the design worries me.  It means that the first time we touch a snapshotted object and need to make a clone, the op is blocked while we go bump a bunch of ref counts.